core/src/main/java/com/linecorp/armeria/common/util/AsyncCloseableSupport.java

```diff
@@ -0,0 +1,211 @@
+
+import com.linecorp.armeria.internal.UnmodifiableFuture;
+
+/**
+ * Provides support for implementing {@link AsyncCloseable}.
+ */
+public final class AsyncCloseableSupport implements AsyncCloseable {
+
+    private static final AtomicIntegerFieldUpdater<AsyncCloseableSupport> closingUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(AsyncCloseableSupport.class, "closing");
+
+    private static final AsyncCloseableSupport CLOSED;
+
+    static {
+        CLOSED = of();
+        CLOSED.closeAsync();
+    }
+
+    /**
+     * Returns a new {@link AsyncCloseableSupport} that will be completed immediately on {@link #close()} or
+     * {@link #closeAsync()}. This method is useful when you don't have any resources to release.
+     * <pre>{@code
+     * > public class MyClass {
+     * >     final AsyncCloseableSupport closeableSupport = AsyncCloseableSupport.of();
+     * >
+     * >     public void doSomething() {
+     * >         if (closeableSupport.isClosing()) {
+     * >             throw new IllegalStateException("Closed already");
+     * >         }
+     * >         ...
+     * >     }
+     * >
+     * >     @Override
+     * >     public CompletableFuture<?> closeFuture() {
+     * >         return closeableSupport.closeFuture();
+     * >     }
+     * >
+     * >     @Override
+     * >     public CompletableFuture<?> closeAsync() {
+     * >         return closeableSupport.closeAsync();
+     * >     }
+     * >
+     * >     @Override
+     * >     public CompletableFuture<?> close() {
+     * >         return closeableSupport.close();
+     * >     }
+     * > }
+     * }</pre>
+     */
+    public static AsyncCloseableSupport of() {
+        return of(f -> f.complete(null));
+    }
+
+    /**
+     * Returns a new {@link AsyncCloseableSupport} which calls the specified {@link Consumer} on
+     * {@link #close()} or {@link #closeAsync()}.
+     * <pre>{@code
+     * > class MyClass implements AutoCloseable {
+     * >     final AsyncCloseableSupport closeableSupport = AsyncCloseableSupport.of(f -> {
+     * >         // Release resources here.
+     * >         ...
+     * >         f.complete(null);
+     * >     });
+     * >
+     * >     @Override
+     * >     public CompletableFuture<?> closeFuture() {
+     * >         return closeableSupport.closeFuture();
+     * >     }
+     * >
+     * >     @Override
+     * >     public CompletableFuture<?> closeAsync() {
+     * >         return closeableSupport.closeAsync();
+     * >     }
+     * >
+     * >     @Override
+     * >     public CompletableFuture<?> close() {
+     * >         return closeableSupport.close();
+     * >     }
+     * > }
+     * }</pre>
+     *
+     * @param closeAction the {@link Consumer} which performs the task that release the resources and
+     *                    completes the given {@link CompletableFuture}.
+     */
+    public static AsyncCloseableSupport of(Consumer<CompletableFuture<?>> closeAction) {
```

Comment:
Thought about it, but `? super` does not make sense here, because the consumer must call `CompletableFuture.complete*()`. A consumer that accepts a `CompletionStage` or `Object` cannot do that.

Reply:
Either way is fine. Sometimes contravariance makes me confused.🤪 
Because we can call `CompletableFuture.complete*()` even it is defined with `Consumer<? super CompletableFuture<?>>`. 🧐
```java
// Compiles
Consumer<? super CompletableFuture<Integer>> consumerSuper = x -> x.complete(10);
```

Reply:
Yeah it compiles, but what I'm saying is this doesn't make sense:

```java
Consumer<CompletionStage<?>> consumer = stage -> ...; // Can't call complete()
AsyncCloseableSupport.of(consumer); // This doesn't make sense. Consumer must call .complete().
```

Reply:
> Consumer must call .complete().

I understood. You may want to constrain consumer implementation. It makes sense to me. 👍